jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -169,20 +169,25 @@ object XGBoost extends Serializable {
       trainingData: Dataset[_],
       params: Map[String, Any],
       round: Int,
-      nWorkers: Int,
+      nWorkers: Int = 0,
       obj: ObjectiveTrait = null,
       eval: EvalTrait = null,
       useExternalMemory: Boolean = false,
       missing: Float = Float.NaN,
       featureCol: String = "features",
       labelCol: String = "label"): XGBoostModel = {
-    require(nWorkers > 0, "you must specify more than 0 workers")
+    val nWorkers_ = if (nWorkers == 0) {
+      val nExecutors = currentActiveExecutorsNum(trainingData.sparkSession.sparkContext)
```

Comment:
You need to times the number of cores per executor as well.

Reply:
Say, we have 5 executors, and each executor owns 4 cores.

Do you suggest to create 20 workers? If I understand correctly, use 5 workers and set `nThread=4` could be better, right? Thanks.

Reply:
In practice, I find that XGBoost will hang forever if `nWorkers > nExecutors`.

@xydrolase 's suggestion sounds reasonable in #1276 :

> It would also work if you set nthread in XGBoost to 4, and configure spark.task.cpus in Spark to 4. Then, use numWorkers = 3, while launching 3 executors.
> The level of parallelism is identical on paper, but in practice might be faster due to less Rabit synchronization etc.

Reply:
by default, xgboost4j will not compile native deps with OpenMP, so set nThread=4 does not make any sense.... @superbobry correct me if I remember wrongly

Reply:
Thanks, all. In fact I have only one question: 

Given: number of executors, tasks and cores
Question: which one is the upper limit for `nWorkers`?

If I understand correctly, number of executors is a safe choice, while number of tasks can get rid of resources idle, right?

Reply:
@CodingCat `xgboost4j` would use OpenMP on Windows and Linux, but on on Mac.